pkg/frontend/transport/handler.go

```diff
@@ -440,13 +442,31 @@ func writeError(w http.ResponseWriter, err error) {
 		}
 	}
 
-	// if the error is an APIError, ensure it gets written as a JSON response
-	if resp, ok := apierror.HTTPResponseFromError(err); ok {
-		_ = httpgrpc.WriteResponse(w, resp)
+	var (
+		res *httpgrpc.HTTPResponse
+		ok  bool
+	)
+
+	// If the error is an APIError, ensure it gets written as a JSON response.
+	// Otherwise, check if there's a response encoded in the gRPC error.
+	res, ok = apierror.HTTPResponseFromError(err)
+	if !ok {
+		res, ok = httpgrpc.HTTPResponseFromError(err)
+	}
+
+	// If we've been able to get the HTTP response from the error, then we send
+	// it with the right status code and response body content.
+	if res != nil {
```

Comment:
Note to reviewers: the logic should have not been changed at all (at least not intentionally). What I'm doing here is exactly what we did before calling `httpgrpc.WriteResponse(w, resp)` and `httpgrpc.WriteError(w, err)`.